### PR TITLE
Add Prefllight logger. Use in RedispatchSiteLocal. Refs #3110

### PR DESCRIPTION
**Pull request by weitzman**

### CODE REVIEW

File: src/Preflight/Preflight.php [hunk 1]
```diff
@@ -60,6 +65,22 @@ public function __construct(Environment $environment, $verify = null, $configLoc
...
+     */
+    public function logger()
+    {
+        return $this->logger;
+    }
+
+    /**
+     * @param PreflightLog $logger
+     */
+    public function setLogger($logger)
```
Comment (greg-1-anderson) [inline src/Preflight/Preflight.php]: You can typehint this too

File: src/Preflight/PreflightLog.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+     */
+    public function setDebug($debug)
+    {
+        $this->debug = $debug;
+    }
+
+    public function log($message)
+    {
+        if ($this->getDebug()) {
+            file_put_contents('php://stderr', ' [preflight] ' . $message . "\n", FILE_APPEND);
```
Comment (greg-1-anderson) [inline src/Preflight/PreflightLog.php]: Maybe `fwrite` is clearer

File: src/Preflight/PreflightArgs.php [hunk 3]
```diff
@@ -227,6 +231,11 @@ public function selectedSite($default = false)
         return $this->get(self::ROOT, $default);
     }
 
+    public function setDebug($value)
+    {
+        $this->set(self::DEBUG, $value);
```
Comment (greg-1-anderson) [inline src/Preflight/PreflightArgs.php]: Sort of wondering if we need to set the debug flag in the pass through args too.

File: src/Preflight/Preflight.php [hunk 4]
```diff
@@ -202,6 +223,11 @@ public function preflight($argv)
         $this->preflightArgs = $this->preflightArgs($argv);
         $this->prepareConfig($this->environment);
 
+        // Setup our early logger.
+        $logger = new PreflightLog();
+        $logger->setDebug($this->preflightArgs->get(PreflightArgs::DEBUG));
+        $this->setLogger($logger);
```
Comment (greg-1-anderson) [inline src/Preflight/Preflight.php]: Usually DI happens from outside the class. Not sure if we want to add another object instantiation to drush_main (or whatever we're calling it) tho

Comment (weitzman) [reply src/Preflight/Preflight.php]: Not sure I follow. This is Preflight and PreflightLog. Two different classes.

File: src/Preflight/PreflightArgs.php (hunk 3 above)
Comment (weitzman) [reply src/Preflight/PreflightArgs.php]: The --debug option is declared in Application.php so nothing more to do, if I understand your comment correctly.

**Comment by weitzman** [comment]
I merged this, after implementing the feedback that I understand. We can do more in follow ups.

File: src/Preflight/Preflight.php (hunk 4 above)
Comment (greg-1-anderson) [reply src/Preflight/Preflight.php]: Yes, `--debug` is defined as a global option; however, if you run `drush status --debug`, it no longer turns on debug mode for most of Drush. The value is only set for preflight.

File: src/Preflight/PreflightArgs.php (hunk 3 above)
Comment (greg-1-anderson) [reply src/Preflight/PreflightArgs.php]: Yes, `--debug` is defined as a global option; however, if you run `drush status --debug`, it no longer turns on debug mode for most of Drush. The value is only set for preflight.

**Comment by weitzman** [comment]
Just pushed a couple commits which fix DI, fixes standard debug log, and uses the new PreflightLog in 3 places.

**Pull request merged**